Add support for SDL3 joystick input driver for Android, iOS, visionOS and Web

Nintorch opened the pull request:
Follow-up to https://github.com/godotengine/godot/pull/106218

Closes https://github.com/godotengine/godot/issues/47656
Closes https://github.com/godotengine/godot/issues/96985

This PR adds support for SDL joystick input for the rest of the platforms Godot Engine supports. Previously it was only available on desktop platforms (Windows, Linux and macOS).

I also decided to refactor `joypad_sdl.cpp` a little bit and add joystick vibration feature to the web platform in this PR :D

I have tested this PR on Android and web, I haven't tested it on iOS and visionOS, but judging by how SDL input was integrated on macOS I hope this approach works for iOS and visionOS too.
I haven't made Android use HIDAPI yet (which was used for new joypad features in my other PR: https://github.com/godotengine/godot/pull/107967 ), I'm not sure I can, because I'm not an Android developer and it was tiring to make SDL's Android Java code work properly in Godot's build system 😅
(EDIT: It can't be done properly: https://github.com/libsdl-org/SDL/commit/e3cf2e479498205cec56906e03800da24e38786b https://github.com/libsdl-org/SDL/blob/main/src/hidapi/android/hid.cpp#L1052 )

This PR also fixes 3 issues on Android (tested with my DualShock 4 on the master branch using [this script](https://gist.github.com/anthonyec/5342fce79b2b7b22ada748df0ad7f7c0)):
<details>
<summary>1. Trigger axes values were not reset after the user stops pressing on them;</summary>

I'm not pressing any buttons on the screenshot:
<img width="516" height="450" alt="image" src="https://github.com/user-attachments/assets/569e5923-5c1d-44f1-a125-338be90f291e" />
Relevant issues https://github.com/godotengine/godot/issues/79263 https://github.com/godotengine/godot/issues/108982

The same situation after my PR:
<img width="399" height="347" alt="image" src="https://github.com/user-attachments/assets/63b0ac9c-8289-4fd1-bb95-3773ec8e667a" />
</details>
<details>
<summary>2. Pressing the left trigger caused the Share button (or JoyButton::BACK) to be pressed as well;</summary>

I'm only pressing the left and right triggers on the screenshot:
<img width="394" height="332" alt="image" src="https://github.com/user-attachments/assets/6f80fa32-88f9-4bec-b970-97f9cb9292a2" />
Pressing Share on the controller didn't trigger anything, but after releasing the left trigger the BACK button stopped being red.
After my PR Godot can properly detect the Share button being pressed:
<img width="431" height="346" alt="image" src="https://github.com/user-attachments/assets/09808e2b-3770-4ffe-b664-2d1d47b5ac0f" />

</details>
<details>
<summary>3. Fingerpring scanners were recognized as joypads.</summary>

Relevant issue: https://github.com/godotengine/godot/issues/47656
Before my PR:
<img width="516" height="450" alt="image" src="https://github.com/user-attachments/assets/569e5923-5c1d-44f1-a125-338be90f291e" />
After my PR:
<img width="432" height="403" alt="image" src="https://github.com/user-attachments/assets/ffdabf13-4092-4e73-ab18-c2bff08fae2f" />

</details>

There's also another Android issue ( https://github.com/godotengine/godot/issues/56181 ), but I can't reproduce it on master with my DualShock 4, so I can't test if it was fixed by my PR.
~~There's another problem, I haven't tested it on master yet but after this PR the joypad triggers are now digital (just 0.0 or 1.0) on web~~ (EDIT: I hope it's now fixed), related SDL issue: https://github.com/libsdl-org/SDL/issues/13051

Old TODO:
- [x] Make a pull request for web joystick vibration for SDL repository ( https://github.com/libsdl-org/SDL/pull/13757 )
- [x] Make a pull request for SDL to fix Android joystick implementation not calling `SDL_ShouldIgnoreJoystick` to ignore unnecessary gamepads ( https://github.com/libsdl-org/SDL/pull/13759 )
- [x] Web: analog triggers
- [x] Web: proper joypad names
- [x] Web: DualSense on Firefox works very badly, I'm not sure it's fixable in Godot, maybe it shouldn't be detected at all?
- [x] SDL: Make a PR for web joysticks to get vendor and product IDs when possible + use them in GUIDs? (see https://stackoverflow.com/questions/66885705/how-to-get-vendor-and-productid-of-joystick-device-with-javascript and https://gist.github.com/Enichan/de1618d2c481523bd0c1123800bfe747 ) ( https://github.com/libsdl-org/SDL/pull/14003 )
- [x] iOS: Undefined haptic symbols problem
- [x] Android: fix crash - https://gist.github.com/bruvzg/9822262842e885170289d40572fc4002

TODO for separate PRs:
- [ ] Windows: Move joypad code to OS_Windows class (it should probably happen after we update to newer version of SDL, since currently SDL needs HWND of a helper window for DirectInput controllers, it's fixed in newer versions of SDL)
- [ ] SDL joypad process and deinitialization macros
- [ ] After https://github.com/godotengine/godot/pull/108214 is merged: `Input.get_joy_info()` fields for new platforms
- [ ] Remove Godot's mapping system code?

TODO for Android:
- [ ] Android: find a way to use SDLSurface.java file's input events to be more stable
- [ ] Check if this issue is fixed on Android: https://github.com/godotengine/godot/issues/73446
- [ ] Check if this PR works on Android: https://github.com/godotengine/godot/pull/109750
- [ ] Check for regressions for fixes in this PR: https://github.com/godotengine/godot/pull/106021 (I currently don't have access to an Xbox controller, so I can't test this myself)

TODO for splitting:
- [ ] PR for Android
- [x] PR for iOS/visionOS ( https://github.com/godotengine/godot/pull/114316 )
- [x] PR for web ( https://github.com/godotengine/godot/pull/114318 )

Nintorch commented:
> Web seems to be extremely buggy:
> - all controllers register as "Standard controller"
> - d-pad doesn't work
> - mapping for face buttons are wrong
> - triggers register only binary states and not as full axis
> - vibration doesn't work
> - there's high input lag, and some presses aren't registered at all
> 
> Android is crashing instantly, in `SDL_InitSubSystem`.
> ```
>       #00 pc 000000000005ba84  /apex/com.android.runtime/lib64/bionic/libc.so (abort+156) (BuildId: 95e892805867a9df9d038d9e274fce41)
>       #01 pc 000000000071f970  /apex/com.android.art/lib64/libart.so (art::Runtime::Abort(char const*)+236) (BuildId: 29de2213ce2448a49b79fc75574aeaf2)
>       #02 pc 0000000000014d1c  /apex/com.android.art/lib64/libbase.so (android::base::SetAborter(std::__1::function<void (char const*)>&&)::$_0::__invoke(char const*)+80) (BuildId: 13d29229303b4debf34deaa44f74426b)
>       #03 pc 00000000000142dc  /apex/com.android.art/lib64/libbase.so (android::base::LogMessage::~LogMessage()+524) (BuildId: 13d29229303b4debf34deaa44f74426b)
>       #04 pc 000000000044fc8c  /apex/com.android.art/lib64/libart.so (art::JavaVMExt::JniAbort(char const*, char const*)+1268) (BuildId: 29de2213ce2448a49b79fc75574aeaf2)
>       #05 pc 000000000044f754  /apex/com.android.art/lib64/libart.so (art::(anonymous namespace)::ScopedCheck::AbortF(char const*, ...) (.__uniq.99033978352804627313491551960229047428)+180) (BuildId: 29de2213ce2448a49b79fc75574aeaf2)
>       #06 pc 00000000004555e0  /apex/com.android.art/lib64/libart.so (art::(anonymous namespace)::ScopedCheck::CheckInstance(art::ScopedObjectAccess&, art::(anonymous namespace)::ScopedCheck::InstanceKind, _jobject*, bool) (.__uniq.99033978352804627313491551960229047428)+1028) (BuildId: 29de2213ce2448a49b79fc75574aeaf2)
>       #07 pc 000000000045a7cc  /apex/com.android.art/lib64/libart.so (art::(anonymous namespace)::ScopedCheck::Check(art::ScopedObjectAccess&, bool, char const*, art::(anonymous namespace)::JniValueType*) (.__uniq.99033978352804627313491551960229047428)+1396) (BuildId: 29de2213ce2448a49b79fc75574aeaf2)
>       #08 pc 0000000000456e34  /apex/com.android.art/lib64/libart.so (art::(anonymous namespace)::CheckJNI::CallMethodV(char const*, _JNIEnv*, _jobject*, _jclass*, _jmethodID*, std::__va_list, art::Primitive::Type, art::InvokeType) (.__uniq.99033978352804627313491551960229047428)+248) (BuildId: 29de2213ce2448a49b79fc75574aeaf2)
>       #09 pc 00000000008a16ac  /apex/com.android.art/lib64/libart.so (art::(anonymous namespace)::CheckJNI::CallStaticVoidMethod(_JNIEnv*, _jclass*, _jmethodID*, ...) (.__uniq.99033978352804627313491551960229047428.llvm.11128407671067317354)+152) (BuildId: 29de2213ce2448a49b79fc75574aeaf2)
>       #10 pc 000000000214ba20  /data/app/~~T6jtuVVyaq82WXRErTJHSw==/com.example.joypads-WM8eoIB890EcL0eVln4mZA==/base.apk!libgodot_android.so (offset 0x3bc000)
>       #11 pc 000000000211d208  /data/app/~~T6jtuVVyaq82WXRErTJHSw==/com.example.joypads-WM8eoIB890EcL0eVln4mZA==/base.apk!libgodot_android.so (offset 0x3bc000)
>       #12 pc 000000000213c928  /data/app/~~T6jtuVVyaq82WXRErTJHSw==/com.example.joypads-WM8eoIB890EcL0eVln4mZA==/base.apk!libgodot_android.so (offset 0x3bc000) (SDL_InitSubSystem+496)
>       #13 pc 000000000211c190  /data/app/~~T6jtuVVyaq82WXRErTJHSw==/com.example.joypads-WM8eoIB890EcL0eVln4mZA==/base.apk!libgodot_android.so (offset 0x3bc000)
>       #14 pc 00000000011fa464  /data/app/~~T6jtuVVyaq82WXRErTJHSw==/com.example.joypads-WM8eoIB890EcL0eVln4mZA==/base.apk!libgodot_android.so (offset 0x3bc000)
>       #15 pc 0000000001259d24  /data/app/~~T6jtuVVyaq82WXRErTJHSw==/com.example.joypads-WM8eoIB890EcL0eVln4mZA==/base.apk!libgodot_android.so (offset 0x3bc000)
>       #16 pc 00000000012160fc  /data/app/~~T6jtuVVyaq82WXRErTJHSw==/com.example.joypads-WM8eoIB890EcL0eVln4mZA==/base.apk!libgodot_android.so (offset 0x3bc000) (Java_org_godotengine_godot_GodotLib_step+108)
>       #17 pc 0000000000386d00  /apex/com.android.art/lib64/libart.so (art_quick_generic_jni_trampoline+144) (BuildId: 29de2213ce2448a49b79fc75574aeaf2)
>       #18 pc 000000000036fc60  /apex/com.android.art/lib64/libart.so (art_quick_invoke_static_stub+640) (BuildId: 29de2213ce2448a49b79fc75574aeaf2)
>       #19 pc 0000000000885da0  /apex/com.android.art/lib64/libart.so (bool art::interpreter::DoCall<false>(art::ArtMethod*, art::Thread*, art::ShadowFrame&, art::Instruction const*, unsigned short, bool, art::JValue*)+1808) (BuildId: 29de2213ce2448a49b79fc75574aeaf2)
>       #20 pc 000000000078fba4  /apex/com.android.art/lib64/libart.so (void art::interpreter::ExecuteSwitchImplCpp<false>(art::interpreter::SwitchImplContext*)+10828) (BuildId: 29de2213ce2448a49b79fc75574aeaf2)
>       #21 pc 0000000000393548  /apex/com.android.art/lib64/libart.so (ExecuteSwitchImplAsm+8) (BuildId: 29de2213ce2448a49b79fc75574aeaf2)
>       #22 pc 0000000000004e1c  [anon:dalvik-classes12.dex extracted in memory from /data/app/~~T6jtuVVyaq82WXRErTJHSw==/com.example.joypads-WM8eoIB890EcL0eVln4mZA==/base.apk] (org.godotengine.godot.gl.GodotRenderer.onDrawFrame+0)
>       #23 pc 0000000000393194  /apex/com.android.art/lib64/libart.so (art::interpreter::ArtInterpreterToInterpreterBridge(art::Thread*, art::CodeItemDataAccessor const&, art::ShadowFrame*, art::JValue*)+548) (BuildId: 29de2213ce2448a49b79fc75574aeaf2)
>       #24 pc 0000000000885d8c  /apex/com.android.art/lib64/libart.so (bool art::interpreter::DoCall<false>(art::ArtMethod*, art::Thread*, art::ShadowFrame&, art::Instruction const*, unsigned short, bool, art::JValue*)+1788) (BuildId: 29de2213ce2448a49b79fc75574aeaf2)
>       #25 pc 000000000078fc7c  /apex/com.android.art/lib64/libart.so (void art::interpreter::ExecuteSwitchImplCpp<false>(art::interpreter::SwitchImplContext*)+11044) (BuildId: 29de2213ce2448a49b79fc75574aeaf2)
>       #26 pc 0000000000393548  /apex/com.android.art/lib64/libart.so (ExecuteSwitchImplAsm+8) (BuildId: 29de2213ce2448a49b79fc75574aeaf2)
>       #27 pc 0000000000003c78  [anon:dalvik-classes12.dex extracted in memory from /data/app/~~T6jtuVVyaq82WXRErTJHSw==/com.example.joypads-WM8eoIB890EcL0eVln4mZA==/base.apk] (org.godotengine.godot.gl.GLSurfaceView$GLThread.guardedRun+0)
>       #28 pc 0000000000393194  /apex/com.android.art/lib64/libart.so (art::interpreter::ArtInterpreterToInterpreterBridge(art::Thread*, art::CodeItemDataAccessor const&, art::ShadowFrame*, art::JValue*)+548) (BuildId: 29de2213ce2448a49b79fc75574aeaf2)
>       #29 pc 0000000000885d8c  /apex/com.android.art/lib64/libart.so (bool art::interpreter::DoCall<false>(art::ArtMethod*, art::Thread*, art::ShadowFrame&, art::Instruction const*, unsigned short, bool, art::JValue*)+1788) (BuildId: 29de2213ce2448a49b79fc75574aeaf2)
>       #30 pc 000000000078fba4  /apex/com.android.art/lib64/libart.so (void art::interpreter::ExecuteSwitchImplCpp<false>(art::interpreter::SwitchImplContext*)+10828) (BuildId: 29de2213ce2448a49b79fc75574aeaf2)
>       #31 pc 0000000000393548  /apex/com.android.art/lib64/libart.so (ExecuteSwitchImplAsm+8) (BuildId: 29de2213ce2448a49b79fc75574aeaf2)
>       #32 pc 0000000000004580  [anon:dalvik-classes12.dex extracted in memory from /data/app/~~T6jtuVVyaq82WXRErTJHSw==/com.example.joypads-WM8eoIB890EcL0eVln4mZA==/base.apk] (org.godotengine.godot.gl.GLSurfaceView$GLThread.run+0)
>       #33 pc 00000000003939c4  /apex/com.android.art/lib64/libart.so (art::interpreter::EnterInterpreterFromEntryPoint(art::Thread*, art::CodeItemDataAccessor const&, art::ShadowFrame*)+636) (BuildId: 29de2213ce2448a49b79fc75574aeaf2)
>       #34 pc 00000000003929b4  /apex/com.android.art/lib64/libart.so (artQuickToInterpreterBridge+988) (BuildId: 29de2213ce2448a49b79fc75574aeaf2)
>       #35 pc 0000000000386e38  /apex/com.android.art/lib64/libart.so (art_quick_to_interpreter_bridge+88) (BuildId: 29de2213ce2448a49b79fc75574aeaf2)
>       #36 pc 000000000036f994  /apex/com.android.art/lib64/libart.so (art_quick_invoke_stub+612) (BuildId: 29de2213ce2448a49b79fc75574aeaf2)
>       #37 pc 000000000035a9bc  /apex/com.android.art/lib64/libart.so (art::ArtMethod::Invoke(art::Thread*, unsigned int*, unsigned int, art::JValue*, char const*)+136) (BuildId: 29de2213ce2448a49b79fc75574aeaf2)
>       #38 pc 00000000004fe240  /apex/com.android.art/lib64/libart.so (art::Thread::CreateCallback(void*)+1004) (BuildId: 29de2213ce2448a49b79fc75574aeaf2)
>       #39 pc 000000000006b768  /apex/com.android.runtime/lib64/bionic/libc.so (__pthread_start(void*)+184) (BuildId: 95e892805867a9df9d038d9e274fce41)
>       #40 pc 000000000005ea18  /apex/com.android.runtime/lib64/bionic/libc.so (__start_thread+64) (BuildId: 95e892805867a9df9d038d9e274fce41)
> ```
> 
> Will check iOS later.

Thank you for testing, bruvzg! That's weird, for me on Windows on web I didn't have problems with dpad, face buttons, vibration and input lag with Dualshock 4 (the other problems did happen, I will try to fix them), and I didn't have crashes on Android. May I ask if there's any other information about the Android crash you're experiencing?

Nintorch commented:
> On iOS, linking fails due to following missing symbols:
> 
> ```
> Undefined symbol: _SDL_SYS_HapticClose
> Undefined symbol: _SDL_SYS_HapticDestroyEffect
> Undefined symbol: _SDL_SYS_HapticInit
> Undefined symbol: _SDL_SYS_HapticQuit
> ```

Oh, that's also strange, because the check in this PR for iOS export template passed.

bruvzg commented:
> Oh, that's also strange, because the check in this PR for iOS export template passed.

CI only build a static library, linking is done when project is exported.

Nintorch commented:
> > May I ask if there's any other information about the Android crash you're experiencing?
> 
> It's crashing on start, tested with "Joypads" demo on Pixel 8 Pro (Android 16) and POCO F3 (LineageOS, Android 15).

Hm, is there any exception thrown or a JNI error shown?

bruvzg commented:
> Hm, is there any exception thrown or a JNI error shown?

```
'JNI DETECTED ERROR IN APPLICATION: CallStaticVoidMethod received NULL jclass
```

Here's full logcat output - https://gist.github.com/bruvzg/9822262842e885170289d40572fc4002

Nintorch commented:
I see, thank you!

bruvzg commented:
A fix for iOS:
```diff
diff --git a/drivers/sdl/SCsub b/drivers/sdl/SCsub
index 3cd3761c4b..8bed438fe6 100644
--- a/drivers/sdl/SCsub
+++ b/drivers/sdl/SCsub
@@ -217,6 +217,7 @@ if env["builtin_sdl"]:
         thirdparty_sources += [
             "core/unix/SDL_appid.c",
             "core/unix/SDL_poll.c",
+            "haptic/dummy/SDL_syshaptic.c",
             "joystick/apple/SDL_mfijoystick.m",
             "thread/pthread/SDL_syscond.c",
             "thread/pthread/SDL_sysmutex.c",
diff --git a/drivers/sdl/SDL_build_config_private.h b/drivers/sdl/SDL_build_config_private.h
index 54c60f32a5..512894146e 100644
--- a/drivers/sdl/SDL_build_config_private.h
+++ b/drivers/sdl/SDL_build_config_private.h
@@ -155,6 +155,7 @@
 #define HAVE_STDIO_H 1
 #define HAVE_LIBC 1
 #define SDL_JOYSTICK_MFI 1
+#define SDL_HAPTIC_DUMMY 1
 #define SDL_TIMER_UNIX 1
 #define SDL_THREAD_PTHREAD 1
```

Nintorch commented:
Thank you! Also, may I ask if the Android crash happened while running the editor build of Godot on your devices or the exported games? I just realized that I only initialize SDL in `platform\android\java\editor\src\main\java\org\godotengine\editor\BaseGodotEditor.kt` (i.e. only in the editor, I assume), and not in `platform\android\java\app\src\com\godot\game\GodotApp.java` (which is used in the exported app, I assume), which is probably the reason the crash happened.

bruvzg commented:
> Also, may I ask if the Android crash happened while running the editor build of Godot on your devices or the exported games?

Exported game, I have not checked the editor.

bruvzg commented:
Tested Android editor, and most stuff seems to be working. Vibration works with XBox One controller, but not with Dual Sense.

Nintorch commented on drivers/sdl/SDL_build_config_private.h:
```diff
@@ -115,19 +115,65 @@
+// Web (Emscripten) defines
+#elif defined(SDL_PLATFORM_EMSCRIPTEN)
+
+#define SDL_PLATFORM_PRIVATE_NAME "Android"
```
```suggestion
#define SDL_PLATFORM_PRIVATE_NAME "Emscripten"
```
I will need to include this fix later

Nintorch commented on drivers/sdl/joypad_sdl.cpp:
```diff
@@ -273,4 +293,27 @@ void JoypadSDL::close_joypad(int p_pad_idx) {
+
+#ifdef ANDROID_ENABLED
+	// Ignore fingerprint sensors
+	if (joy_name_lower.begins_with("uinput-")) {
```
This has been fixed in upstream SDL, so I will need to remove this fix here

Nintorch committed:
SDL joystick input for Android, iOS, visionOS, Web

This commit adds support for SDL joystick input for the rest of the platforms Godot Engine supports. Previously it was only available on desktop platforms (Windows, Linux and macOS). This commit also adds joystick vibration feature to the web platform in browsers that support this feature.

m4gr3d commented on .github/workflows/runner.yml:
Are those changes intended?

Nintorch replied:
No, I wanted to test the web joysticks more easily. I will revert the changes later!

Nintorch commented on platform/android/detect.py:
```diff
@@ -246,3 +246,10 @@ def configure(env: "SConsEnvironment"):
+        if env["builtin_sdl"]:
+            env.Append(CPPDEFINES=["SDL_ENABLED"])
+        else:
+            print_warning("`builtin_sdl` was explicitly disabled. Disabling SDL input driver support.")
```
Yes, thanks for the change!

Nintorch commented on platform/android/os_android.h:
```diff
@@ -40,6 +40,7 @@ class GodotJavaWrapper;
 class GodotIOJavaWrapper;
 
 struct ANativeWindow;
+class JoypadSDL;
```
Since it's only a class declaration, I think it's okay since we only need a pointer to this class inside the `OS_Android` class, that's also how it's done in other platforms.

Nintorch commented:
Thank you for your review, m4gr3d! I'm not sure I will be able to work on this PR for a while, but when I do, I will make sure I include your requested changes!

Nintorch commented:
Another idea: when I have enough free time for this PR, should it also be split just like https://github.com/godotengine/godot/pull/107967 so it can be reviewed more easily, or these platforms should be merged together in one PR like https://github.com/godotengine/godot/pull/106218 ?

Nintorch commented:
I have split this PR into 2 PRs for iOS, visionOS, and Web. Android still needs some more work done, particularly I think some of the Java input code has to be separated from the code that deals with Activity and View classes (I don't think Godot can use SDL's Activity and View classes), I might make a PR for that later if it's okay with SDL's developers.

Nintorch commented:
These PR might be useful for us to implement SDL for Android in the future:
- https://github.com/libsdl-org/SDL/pull/15199
- https://github.com/mavlink/qgroundcontrol/pull/13885
